python/cuda_cooperative/cuda/cooperative/experimental/_types.py

```diff
@@ -148,15 +148,27 @@ def __repr__(self) -> str:
         return f'Pointer(dtype={self.value_dtype}, out={self.is_output})'
 
     def cpp_decl(self, name):
-        return numba_type_to_cpp(self.value_type) + '* ' + name
+        return numba_type_to_cpp(self.value_dtype) + '* ' + name
 
     def dtype(self):
-        return numba.types.Array(self.value_dtype, 1, 'C')
+        return numba.types.Array(self.value_dtype, 1, 'A')
```

Comment:
I made this change because I was [getting failures in `array_to_array` in Numba](https://github.com/numba/numba/blob/release0.60/numba/np/arrayobj.py#L6457-L6461), which gets called when there's an implicit cast between array types. It's unhappy if you're casting to an array type that doesn't have the `'A'` layout. The `'A'` layout means the array could be in any layout, so this makes sense; you can cast a `'C'` (row-major) or `'F'` (column-major) array to `'A'`, but not vice versa. It's akin to being able to cast anything to `void*` implicitly, but not explicitly.

Perhaps Graham Markall can provide some insight.